File: front-end/src/app/domain/models/game/action/impl/card/JailBreakCardPurchaseAction.ts
import {GameAction} from "../../GameAction";
import {Player} from "../../../Player";
import {JailBreakCard} from "../../../card/impl/JailBreakCard";
import {AssetPurchaseAction} from "../money/AssetPurchaseAction";

/**
 * TODO: Add JsDoc
 *
 * @author Hamza el Haouti
 */
export class JailBreakCardPurchaseAction extends AssetPurchaseAction {
  public constructor(
    id: number,
    newOwner: Player,
    previousOwner: Player,
    performedAt: Date,
    purchasePrice: number,
    card: JailBreakCard
  ) {
    super(JailBreakCardPurchaseAction.name, id, newOwner, previousOwner, performedAt, purchasePrice);
    this._card = card;
  }

  private _card: JailBreakCard;

  public get card(): JailBreakCard {
    return this._card;
  }

  public static trueCopy(action: JailBreakCardPurchaseAction): GameAction | null {
    return new JailBreakCardPurchaseAction(
      action.id,
      Player.trueCopy(action.actor),
      Player.trueCopy(action.recipient),
      new Date(action.performedAt),
      action.amount,
      JailBreakCard.trueCopy(action.card)
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(), {
        "card": this.card.toJson(),
      }
    );
  }

  public toString(): string {
    return `${this.actor.user.username} purchased a ${this.card.toString()} from ${this.recipient.user.username}`;
  }

}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/card/JailBreakCardPurchaseAction.java
package nl.hva.backend.domain.models.game.action.impl.card;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.impl.money.moneyExchangeImpl.AssetPurchaseAction;
import nl.hva.backend.domain.models.game.card.impl.JailBreakCard;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;

@Entity
@Getter
@Setter
@NoArgsConstructor
public class JailBreakCardPurchaseAction extends AssetPurchaseAction {

    @ManyToOne
    private JailBreakCard card;

    public JailBreakCardPurchaseAction(
            Player buyer,
            Player previousOwner,
            double purchasePrice,
            JailBreakCard card
    ) {
        super(buyer, previousOwner, purchasePrice);
        this.card = card;
    }

}